hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/mongodb/FieldResolver.java

```diff
@@ -0,0 +1,226 @@
+                    throw new IllegalArgumentException("Could not load column class " + documentField.columnType, e);
+                }
+            }
+        } else {
+            for (MappingField f : userFields) {
+                String prefixIfStream = stream ? "fullDocument." : "";
+                String nameInMongo = f.externalName() == null ? prefixIfStream + f.name() : f.externalName();
+
+                DocumentField documentField = dbFields.get(nameInMongo);
+                if (documentField == null) {
+                    throw new IllegalArgumentException("Could not resolve field with name " + nameInMongo);
+                }
+                MappingField mappingField = new MappingField(f.name(), f.type(), nameInMongo);
+                mappingField.setPrimaryKey(isId(documentField.columnName, stream));
+                validateType(f, documentField);
+                resolvedFields.add(mappingField);
+            }
+        }
+        return resolvedFields;
+    }
+
+    boolean isId(String nameInMongo, boolean stream) {
+        if (stream) {
+            return "fullDocument._id".equalsIgnoreCase(nameInMongo);
+        }
+        return "_id".equalsIgnoreCase(nameInMongo);
+    }
+
+    private QueryDataType resolveType(BsonType columnType) {
+        switch (columnType) {
+            case INT32: return QueryDataType.INT;
+            case INT64: return QueryDataType.BIGINT;
+            case DOUBLE: return QueryDataType.DOUBLE;
+            case BOOLEAN: return QueryDataType.BOOLEAN;
+            case TIMESTAMP:
+            case DATE_TIME:
+                return QueryDataType.TIMESTAMP;
+            case STRING:
+            case OBJECT_ID:
+                return VARCHAR;
+            case DECIMAL128: return QueryDataType.DECIMAL;
+            default:  throw new UnsupportedOperationException("Cannot resolve type for BSON type " + columnType);
+        }
+    }
+
+    private void validateType(MappingField field, DocumentField documentField) {
+        QueryDataType type = resolveType(documentField.columnType);
+        if (!field.type().equals(type) && !type.getConverter().canConvertTo(field.type().getTypeFamily())) {
+            throw new IllegalStateException("Type " + field.type().getTypeFamily() + " of field " + field.name()
+                    + " does not match db type " + type.getTypeFamily());
+        }
+    }
+
+    Map<String, DocumentField> readFields(Map<String, String> options, boolean stream) {
+        Map<String, DocumentField> fields = new HashMap<>();
+        try (MongoClient client = connect(options)) {
+            String databaseName = requireNonNull(options.get(DATABASE_NAME_OPTION),
+                    DATABASE_NAME_OPTION + " option must be provided");
+            String collectionName = options.get(COLLECTION_NAME_OPTION);
+
+            MongoDatabase database = client.getDatabase(databaseName);
+            List<Document> collections = database.listCollections()
```

Comment:
You can probably use database.getCollection() method by passing the collection name instead of listCollections() method

Reply:
Yes, but actually no ;)

In general you are right, but getCollection's result won't have the information about collection options - and I want to get validator from collection's options